### ISSUE TITLE
[b] merge_gds output fails checker in non-default case

### ISSUE THREAD

**Issue by aofarrel** [DataBiosphere/analysis_pipeline_WDL#20]
Although it passes the check_merged_gds step, the nondefaults case failed the checker workflow in the most recent run on Terra and on local. The checker workflow truth files were generated on SB.

Local output file: 624,440 bytes, b11effbaeab5380c27862e4fc3b7c828
Terra output file: 624,440 bytes, b11effbaeab5380c27862e4fc3b7c828
Nondefaults truth: 615,984 bytes

Ruled out:
* Just mixed up the default and nondefault truth files
  * logs confirm the file localized was from nondefault folder in google bucket
  * truth file in nondefault gs bucket is 615,984 bytes exactly matching size of output file for the nondefault run on SB
  * the truth file for the default run is like 100,000 bytes larger due to having less limitations; the difference between the two is not trivial
* ~~Something went wrong in the ld pruning or subset steps~~
  * ~~checker also meticulously checks the output of subset_gds against truth files, all of them passed, on both Terra and local~~
  * actually, something went wrong during those checks, but it's not clear why. see below.

**Comment by aofarrel** [comment]
SB config for nondefault merge step:
> gds_file "1KG_phase3_subset_subset_chr .gds"
> merged_gds_file "includePCA_hg19_10.1win_0.3r_0.05MAF_0.02missing.gds"
> chromosomes "1 10 11 12 13 14 15 16 17 18 19 2 20 21 22 3 4 5 6 7 8 9"

Terra config file for nondefault merge step:

> gds_file "1KG_phase3_subset_subset_chr .gds"
> merged_gds_file "includePCA_hg19_10.1win_0.3r_0.05MAF_0.02missing.gds"
> chromosomes "1 10 11 12 13 14 15 16 17 18 19 2 20 21 22 3 4 5 6 7 8 9"
> 

Only difference is add'l newline at the end of the Terra one (ie it's four lines instead of three).

**Comment by aofarrel** [comment]
I have a run from earlier this month on Terra that passed this test, but something seems fishy. In the defaults case, we can see the OK clearly.

> 2021/05/12 00:46:29 Running user action: docker run -v /mnt/local-disk:/cromwell_root --entrypoint=/bin/bash python@sha256:eef8e3656385c56c95f08f004c0514a1ad96f3a3791a7d6f6be04554f2ab51a8 /cromwell_root/script
> Information about these truth files:
> Container:	uwgac/topmed-master:2.10.0
> Platform:	SB
> Date updated:	May 10, 2021
> The container version refers to the container used in applicable tasks in the WDL and is the important value here.
> If container versions are equivalent, there should be no difference in GDS output between a local run and a run on Terra.
> test file: merged.gds
> truth file(s): /cromwell_root/topmed_workflow_testing/UWGAC_WDL/checker/b_ldpruning/merged.gds
> /cromwell_root/topmed_workflow_testing/UWGAC_WDL/checker/b_ldpruning/merged.gds: OK
> 2021/05/12 00:46:31 Starting delocalization.

But in the nondefaults case, that's not seen. The test file also has the wrong name (although I do remember setting up the nondefaults to give that name was the last thing I added, so the variables that actually matter were *probably* set...)

> 2021/05/12 00:46:33 Running user action: docker run -v /mnt/local-disk:/cromwell_root --entrypoint=/bin/bash python@sha256:eef8e3656385c56c95f08f004c0514a1ad96f3a3791a7d6f6be04554f2ab51a8 /cromwell_root/script
> Information about these truth files:
> Container:	uwgac/topmed-master:2.10.0
> Platform:	SB
> Date updated:	May 10, 2021
> The container version refers to the container used in applicable tasks in the WDL and is the important value here.
> If container versions are equivalent, there should be no difference in GDS output between a local run and a run on Terra.
> test file: merged.gds
> truth file(s): /cromwell_root/topmed_workflow_testing/UWGAC_WDL/checker/b_ldpruning/nondefault/includePCA_hg19_10.1win_0.3r_0.05MAF_0.02missing.gds
> 2021/05/12 00:46:35 Starting delocalization.
> 2021/05/12 00:46:36 Delocalization script execution started...

TL;DR this might have been broken from the start but it's only clear now.

**Comment by aofarrel** [comment]
> vcf <- seqGDS2VCF("UWGAC_WDL_checker_b_ldpruning_nondefault_includePCA_hg19_10.1win_0.3r_0.05MAF_0.02missing.gds", "truth.vcf")
Thu May 20 09:56:59 2021
VCF Export: truth.vcf
    **1,126 samples, 2,668 variants**
    INFO Field: <none>
    FORMAT Field: <none>
    output to a general gzip file
[==================================================] 100%, completed, 0s
Thu May 20 09:56:59 2021    Done.
> vcf2 <- seqGDS2VCF("TERRA_BAD_MERGE_NONDEFAULTS.gds", "test.vcf")
Thu May 20 10:56:45 2021
VCF Export: test.vcf
    **1,126 samples, 2,703 variants**
    INFO Field: <none>
    FORMAT Field: <none>
    output to a general gzip file
[==================================================] 100%, completed, 1s
Thu May 20 10:56:46 2021    Done.

So it's not just a rearrangement. There's actually some extra variants.

**Comment by aofarrel** [comment]
VCF files are identical for first 1270 lines, then the test file has...
17	40572793	rs57307201	G	A	100	PASS	.	GT
17	41950577	rs1072235	C	T	100	PASS	.	GT
17	42074901	rs3964442	G	A	100	PASS	.	GT
17	43735258	rs1635301	T	G	100	PASS	.	GT
17	43902842	rs41280116	C	T	100	PASS	.	GT

We go back into total alignment, until 1627, where the test file has...
2	129966547	rs6724113	T	C	100	PASS	.	GT
2	130041099	rs837850	G	A	100	PASS	.	GT
2	130664707	rs34066555	G	C	100	PASS	.	GT
2	133950661	rs2082627	G	A	100	PASS	.	GT
2	138451475	rs13415092	C	T	100	PASS	.	GT
2	138597458	rs1455993	A	T	100	PASS	.	GT
2	138932994	rs12986584	C	A	100	PASS	.	GT

After that we return to alignment. At line 1640 things get weird...
Test file:
2	150363247	rs7370844	C	T	100	PASS	.	GT

Truth file:
2	150351196	rs4368294	G	A	100	PASS	.	GT

The truth file completely lacks rs7370844 and the test file lacks rs4368294.

There's a few more differences but really the main insight here is that, yes indeed, actual variants are different. Considering the LD pruning parts all passed, this is bizarre.

**Comment by aofarrel** [comment]
One *very minor* difference between the CWL and WDL is that the CWL (apparently by mistake) does not use the specified output prefix. In other words...
CWL:
prefix_ld_pruning.RData --> subset.gds --> prefix_merged.gds

WDL:
prefix_ld_pruning.RData --> prefix_subset.gds --> prefix_merged.gds

Surely that is not it. Surely. _Surely_.

**Comment by aofarrel** [comment]
Next steps:
~~1. Regenerate truth file on SB~~ it's still 615,984 bytes
2. Double-triple-check the config files and inputs across SB and WDL
3. Rerun checker without the extra newline in the merge config file
4. Rerun checker without the prefix in the subset step

**Comment by aofarrel** [comment]
CWL SB ld_pruning.config for scatter 19 (chr 6)
> gds_file "/sbgenomics/Projects/bb891fb3-b717-4069-9564-74f1e36442a7/1KG_phase3_subset_chr6.gds"
> genome_build "hg19"
> ld_r_threshold 0.3
> ld_win_size 10.1
> maf_threshold 0.05
> missing_threshold 0.02
> out_file "includePCA_hg19_10.1win_0.3r_0.05MAF_0.02missing_pruned_variants_chr6.RData"
> \n

WDL local ld_pruning.config for shard 18 (chr 6)

> gds_file "/bark-bark/checker/a948c1c6-3443-40fb-86f5-a6dcef344424/call-prune_nondefaults/shard-18/inputs/1648365416/1KG_phase3_subset_chr6.gds"
**> exclude_pca_corr false**
> genome_build "hg19"
> ld_r_threshold 0.3
> ld_win_size 10.1
> maf_threshold 0.05
> missing_threshold 0.02
> out_file "includePCA_hg19_10.1win_0.3r_0.05MAF_0.02missing_pruned_variants_chr6.RData"
> \n

...ah.

**Comment by aofarrel** [comment]
MD5 (test_local_chr6.RData) = 01654e987fa1cb5b9d30681bf5825783
MD5 (truth_sb_chr6.RData) = 218b73d783fe72a311efe9a173f68d53
MD5 (truth_sb_subset_chr6.gds) = d58fb9c749c6db2ebbef7b969923e7b0
MD5 (truth_topmedbucket_subset_chr6.gds) = d58fb9c749c6db2ebbef7b969923e7b0
MD5 (test_subset_chr6.gds) = 4e909626b8a8dc730604a7f7df25f9d3
MD5 of local checker run chr6 gds: 4e909626b8a8dc730604a7f7df25f9d3
Stdout of local checker run chr6 gds: 
/bark-bark/checker/a948c1c6-3443-40fb-86f5-a6dcef344424/call-md5sum_wfB_nondefaults/shard-18/inputs/-3847984/1KG_phase3_subset_subset_chr6.gds: FAILED

But that doesn't explain why the Terra logs are showing OK in their logs. They are in CallCache, though... but call cacheing should not be able to cause this sort of problem.

**Comment by aofarrel** [comment]
Currently rerunning the checker on Terra with call cacheing enabled, [albeit as another workflow](https://support.terra.bio/hc/en-us/community/posts/360077995811-Cannot-update-workflow-if-current-snapshot-active-on-Terra-becomes-invalid) instead of another snapshot.

If a call cacheing run succeeds in the nondefault subset md5, and a non call cached run fails in the nondefault subset md5 (both should fail at merge check regardless), then this could be a call cacheing bug.
If call cached run fails, we'll chalk up that successful run as a mysterious fluke.

**Comment by aofarrel** [comment]
The non-call cached run... had a syntax error in the subset section? Why did the non-call-cache one not have this?
+ python
Traceback (most recent call last):
  File "<stdin>", line 42, in <module>
  File "<stdin>", line 10, in py_rootPlusChr
AttributeError: type object 'file' has no attribute 'split'

**Comment by aofarrel** [comment]
The call cached and non-cached runs both succeeded on the subset check and failed on the last check. There's gotta be something wrong with the data going into that subset check.

**Comment by aofarrel** [comment]
Defaults md5sum:
```
	scatter(wfB_gds_test in subset_defaults.subset_output) {
		call md5sum as md5sum_wfB_defaults {
			input:
				gds_test = wfB_gds_test,
				gds_truth = wfB_truth_defaults_subset,
				truth_info = wfB_truth_defaults_info
		}
	}
```
Non-defaults md5sum:
```
	scatter(wfB_gds_test in subset_nondefaults.subset_output) {
		call md5sum as md5sum_wfB_nondefaults {
			input:
				gds_test = wfB_gds_test,
				gds_truth = wfB_truth_nondefaults_subset,
				truth_info = wfB_truth_nondefaults_info
		}
	}
```

Even if wfB_gds_test is somehow getting trumped by the earlier call of the same name, the truth files differ.

MD5 (nondefault_chr6_truth.gds) = d58fb9c749c6db2ebbef7b969923e7b0
MD5 (default_chr6_truth.gds) = bba0ead8cdbc35059fbe3e1f75097151
What's getting written to sum.txt for chr6 in local run: 4e909626b8a8dc730604a7f7df25f9d3  /bark-bark/checker/a948c1c6-3443-40fb-86f5-a6dcef344424/call-md5sum_wfB_nondefaults/shard-18/inputs/1103838408/1KG_phase3_subset_subset_chr6.gds

...is the test file is getting written instead of the truth file so it just checks against itself

**Comment by aofarrel** [comment]
Local test:
* no call cacheing
* concurrent-job-limit = 1
* set -eux -o pipefail
* md5 subset task should occur before md5 merged task

shard-11 of md5 subset task: md5sum results in a mismatch between chr2 truth and test files, as expected

However, the pipeline seems to have just kept going. The error Cromwell reported on the command line was in the md5 merge task. This doesn't seem to make sense; if an md5 mismatch causes a pipeline error in the md5 merge task, it should also be causing a pipeline failure in the md5 subset task.

**Comment by aofarrel** [comment]
Moving this to a new task as the mystery of why the checker was failing (exclude_pca_corr) has bene solved.
